OLM-2608: draft bootstrapping (#17)
grokspawn/cool-catalog: README.md

5. push changes

## Detailed HOWTO

### Actors and Terms
- [`File-Based Catalog`](https://olm.operatorframework.io/docs/reference/file-based-catalogs/) (FBC) is the declarative expression of operators and their relationships with other operators, other versions of themselves.
- [`Veneers`](https://olm.operatorframework.io/docs/reference/veneers/) are a general class of objects which can provide a simplified interaction with FBC.
- `Operator Author` is the role related to expressing an individual operator versions, channels, properties, etc. in a destination catalog.
- `Catalog` is the FBC-based catalog composed of the FBC contributions of one or more operators.
- `Catalog Owner` is the role related to integrating Operator Authors' catalog contributions.
- `Catalog Contribution` is the FBC that a Catalog Ownder receives from an Operator Author.  The format needs to be negotiated with the Operator Author.
For this example, the Catalog Owner receives contributions of a single, unversioned directory named after the operator which contain all FBC files, e.g.:

```tree
catalog
└── testoperator
    ├── .indexignore
    ├── OWNERS

>     └── catalog.yaml

Reviewer: might be helpful to include a testoperator2 just so that we're being very explicit about each author's territory and what's out of bounds for changes by the individual authors?

```

### Lifecycle

> This repository models a FBC catalog which receives contributions from one or more Operator Authors, merges and organizes those contributions. This repository has pre-configured GitHub actions to validate contributing authors, validate FBC contributions, auto-merge contribution PRs, and push updated catalog images capable of serving as [Catalog Sources](https://olm.operatorframework.io/docs/concepts/crds/catalogsource/).

Reviewer: ```suggestion
This repository models an operator catalog which receives contributions from one or more Operator Authors, merges and organizes those contributions. This repository has pre-configured GitHub actions to validate contributing authors, validate contributed content, auto-merge contribution PRs, and push updated catalog images capable of serving as [Catalog Sources](https://olm.operatorframework.io/docs/concepts/crds/catalogsource/).
```

```mermaid
%%{init: {'securityLevel': 'strict', 'theme':'forest'}}%%
sequenceDiagram
autonumber
participant CatalogOwner
participant Catalog
participant GHAction
participant ImageRepo
CatalogOwner--xCatalog: fork repo
CatalogOwner--xCatalog: adjust CI
CatalogOwner--xCatalog: push updates to remote
CatalogOwner->>+Catalog: Approves PR, merges (auto-merge trivial case)
activate CatalogOwner
activate Catalog
Catalog->>+GHAction: Integrate contribution to catalog
activate GHAction
GHAction->>GHAction: validate contribution's Operator Author 
GHAction->>GHAction: generate image contiaining updated catalog
GHAction->>GHAction: validate catalog in image 
GHAction->>+ ImageRepo: push catalog image to repo
activate ImageRepo
deactivate ImageRepo
deactivate GHAction
deactivate Catalog
deactivate CatalogOwner  

> ```

Reviewer: just realized this doesn't make much sense while reviewing in a PR 😄 

Is it possible to add an image as a comment?

Author: branch view in github is a great way to preview the doc as it would actually appear: https://github.com/grokspawn/cool-catalog/tree/bootstrapping

1. Fork and Clone Remote Repository  

> IMPORTANT! Failure to customize could result in unwanted contributions.

Reviewer: This should be prevented by the Catalog authors by merge branch protection right?

Author: That's the premise, yes.  As-is, the repository would accept (and auto-merge) anything that a grokspawn/example-operator-index was configured to point to.

2. Determine Storage Strategy and Adjust CI

> The default configuration for this repository is a collection of operator-titled directories which individually and collectively validate successfully via `opm validate`. The storage strategy must be communicated to any contributing Operator Authors.

Reviewer: do we wanna link to https://olm.operatorframework.io/docs/reference/file-based-catalogs/#cli somewhere in this doc? Also realized that doc itself could benefit from a link to the opm releases url

> Customize CI workflows in `.github/workflows/` to suit your authorization, validation, organization, and publication needs.

Reviewer: `.github/workflows/` can be a link to the actual file right?

3. Push Changes to Remote

> Once finishing customization, push changes to the remote repo to ready it to receive contributions.

Reviewer: is it valuable to walk through an actual example while listing these steps? It's not uncommon for contributors to be non-devs (i.e folks without solid git knowledge to understand what exactly needs to be done)

Author: Right now the whole repo /is/ a working example, along with the example-operator-index repo.  The landscape of what is stored in a catalog and what folks want to validate is vast.  I'm not against iterating to add good details for prominent use-cases we identify, but right now I don't have that info.

4. Catalog Contribution Approval
The default configuration of this repo will auto-approve an Operator Author's contribution if:

    1. the Operator Author is listed in the CODEOWNERS file
    2. the Catalog Contribution passes validation.  This is currently just `opm validate`.

>     3. the Catalog Contribution cam be automerged without conflict.  

Reviewer: ```suggestion
    3. the Catalog Contribution can be automerged without conflict.  
```
